pkg/provider/conversions.go

```diff
@@ -25,79 +25,82 @@ import (
+	}
 
-			if requiredEmailDomain != "" && strings.EqualFold(userDomain, requiredEmailDomain) {
-				foundSeeds = append(foundSeeds, seed)
-				foundDatacenters = append(foundDatacenters, datacenter)
-				continue iterateOverSeeds
-			}
+	if count := len(matchingDatacenters); count > 1 {
+		return nil, nil, errors.New(http.StatusInternalServerError, fmt.Sprintf("expected to find exactly one datacenter with name %q, got %d", datacenterName, count))
+	}
 
-			for _, whitelistedDomain := range requiredEmailDomains {
-				if whitelistedDomain != "" && strings.EqualFold(userDomain, whitelistedDomain) {
-					foundSeeds = append(foundSeeds, seed)
-					foundDatacenters = append(foundDatacenters, datacenter)
-					continue iterateOverSeeds
-				}
-			}
+	matchingDatacenter := matchingDatacenters[0]
+	matchingSeed := matchingSeeds[0]
+
+	if !userInfo.IsAdmin {
+		emailDomain, err := getUserEmailDomain(userInfo)
+		if err != nil {
+			return nil, nil, errors.New(http.StatusInternalServerError, fmt.Sprintf("invalid email domain: %v", err))
 		}
-	}
 
-	if len(foundDatacenters) == 0 {
-		return nil, nil, errors.New(http.StatusNotFound, fmt.Sprintf("datacenter %q not found", datacenterName))
-	}
-	if n := len(foundDatacenters); n > 1 {
-		return nil, nil, fmt.Errorf("expected to find exactly one datacenter with name %q, got %d", datacenterName, n)
+		if !canAccessDatacenter(matchingDatacenter, emailDomain) {
+			return nil, nil, errors.New(http.StatusForbidden, fmt.Sprintf("cannot access %s datacenter due to email domain requirements", datacenterName))
+		}
 	}
 
-	return foundSeeds[0], &foundDatacenters[0], nil
+	return matchingSeed, &matchingDatacenter, nil
 }
 
-func skipFilterByDomain(userInfo *UserInfo, datacenter kubermaticv1.Datacenter) bool {
-	requiredEmailDomain := datacenter.Spec.RequiredEmailDomain
-	requiredEmailDomains := datacenter.Spec.RequiredEmailDomains
+// canAccessDatacenter returns information if user with provided email domain can access given datacenter
+// based on the required email domains that are set on it.
+func canAccessDatacenter(dc kubermaticv1.Datacenter, emailDomain string) bool {
+	// Return false if required email domain is set, but it is different from user email domain.
+	if dc.Spec.RequiredEmailDomain != "" && !strings.EqualFold(emailDomain, dc.Spec.RequiredEmailDomain) {
+		return false
+	}
 
-	if userInfo.IsAdmin {
-		return true
+	// Return false if required email domains are set, but all of them are different from user email domain.
+	if len(dc.Spec.RequiredEmailDomains) > 0 {
+		isMatching := false
+		for _, domain := range dc.Spec.RequiredEmailDomains {
+			if domain != "" && strings.EqualFold(emailDomain, domain) {
+				isMatching = true
+				break
+			}
+		}
+		if !isMatching {
+			return false
+		}
 	}
-	if requiredEmailDomain == "" && len(requiredEmailDomains) == 0 {
-		return true
+
+	return true
+}
+
+func getUserEmailDomain(userInfo *UserInfo) (string, error) {
+	split := strings.Split(userInfo.Email, "@")
```

Comment:
The spec for valid email addresses permits multiple "@" signs in an email address. You want to use the last one as the domain part, not the first one: https://play.golang.org/p/cg4RqZADLml

Reply:
This is code moved from the old version, but sure, I'll fix it.